pl.Expr.any() produces wrong results in windows functions

Linux Mint 20.3
python 3.10.2
polars 0.13.3

#### Describe your bug.
When the pl.Expr.any() expression is used over a window function, it produces incorrect results.

**Caution**: the results are **non-deterministic**.  Using the same starting data, the pl.Expr.any() expression produces different answers with each run (and this is not a result of using a random number generator to create the initial dataset).
#### What are the steps to reproduce the behavior?
```python
import random

import polars as pl

# Generate an initial dataset with two rows per row_id and a random selection of True/False values for each row
df = (
    pl.DataFrame(
        {
            "row_id": list(range(0, 500)) * 2,
            "boolvar": [random.getrandbits(1) for x in range(0, 1000)],
        }
    )
    .sort(["row_id"])
    .with_column(pl.col("boolvar").cast(pl.datatypes.Boolean).keep_name())
)
print(df)
```
```
┌────────┬─────────┐
│ row_id ┆ boolvar │
│ ---    ┆ ---     │
│ i64    ┆ bool    │
╞════════╪═════════╡
│ 0      ┆ false   │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ 0      ┆ true    │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ 1      ┆ false   │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ 1      ┆ true    │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ ...    ┆ ...     │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ 498    ┆ false   │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ 498    ┆ true    │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ 499    ┆ false   │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ 499    ┆ false   │
└────────┴─────────┘
```

Now, run a windowed pl.Expr.any() expression -- twice -- on the **same** data...
```python
pl.testing.assert_frame_equal(
    df.with_column(pl.any("boolvar").over("row_id").alias("result")),
    df.with_column(pl.any("boolvar").over("row_id").alias("result")),
)
```
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/corey/.virtualenvs/test_project3.10/lib/python3.10/site-packages/polars/testing.py", line 99, in assert_frame_equal
    _assert_series_inner(
  File "/home/corey/.virtualenvs/test_project3.10/lib/python3.10/site-packages/polars/testing.py", line 183, in _assert_series_inner
    raise_assert_detail(
  File "/home/corey/.virtualenvs/test_project3.10/lib/python3.10/site-packages/polars/testing.py", line 209, in raise_assert_detail
    raise AssertionError(msg)
AssertionError: DataFrame are different
```
#### What is the actual behavior?
Let's take a closer look at any pairs of row_id's for which the windowed pl.Expr.any() expression produced different values:
```python
df = df.with_column(pl.any("boolvar").over("row_id").alias("result1"))
df = df.with_column(pl.any("boolvar").over("row_id").alias("result2"))

diffs = df.filter((pl.col("result1") != pl.col("result2")))

df.join(diffs.select(pl.col("row_id").unique()), on="row_id")
```

One my *first* run using the generated data, I receive this result:
```
shape: (62, 4)
┌────────┬─────────┬─────────┬─────────┐
│ row_id ┆ boolvar ┆ result1 ┆ result2 │
│ ---    ┆ ---     ┆ ---     ┆ ---     │
│ i64    ┆ bool    ┆ bool    ┆ bool    │
╞════════╪═════════╪═════════╪═════════╡
│ 72     ┆ false   ┆ true    ┆ true    │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ 72     ┆ true    ┆ false   ┆ true    │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ 104    ┆ false   ┆ true    ┆ true    │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ 104    ┆ true    ┆ false   ┆ true    │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ ...    ┆ ...     ┆ ...     ┆ ...     │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ 482    ┆ true    ┆ false   ┆ true    │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ 482    ┆ false   ┆ true    ┆ true    │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ 491    ┆ true    ┆ true    ┆ false   │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ 491    ┆ false   ┆ true    ┆ true    │
└────────┴─────────┴─────────┴─────────┘
```
Note how for some row_id's, the windowed pl.Expr.any() expression does not produce a consistent value for both records with the same row_id.

Let's run just the windowed pl.Expr.any() expressions again on the **same** data (that is, **without** regenerating any new randomized boolean values).
```python
df.drop_in_place('result1')
df.drop_in_place('result2')

df = df.with_column(pl.any("boolvar").over("row_id").alias("result1"))
df = df.with_column(pl.any("boolvar").over("row_id").alias("result2"))

diffs = df.filter((pl.col("result1") != pl.col("result2")))

df.join(diffs.select(pl.col("row_id").unique()), on="row_id")
```
Now I receive different results:
```
shape: (22, 4)
┌────────┬─────────┬─────────┬─────────┐
│ row_id ┆ boolvar ┆ result1 ┆ result2 │
│ ---    ┆ ---     ┆ ---     ┆ ---     │
│ i64    ┆ bool    ┆ bool    ┆ bool    │
╞════════╪═════════╪═════════╪═════════╡
│ 67     ┆ false   ┆ true    ┆ false   │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ 67     ┆ true    ┆ true    ┆ true    │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ 131    ┆ false   ┆ true    ┆ false   │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ 131    ┆ true    ┆ true    ┆ true    │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ ...    ┆ ...     ┆ ...     ┆ ...     │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ 450    ┆ true    ┆ true    ┆ true    │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ 450    ┆ false   ┆ true    ┆ false   │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ 482    ┆ true    ┆ false   ┆ false   │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ 482    ┆ false   ┆ true    ┆ false   │
└────────┴─────────┴─────────┴─────────┘
```
Note how: 
* the number of records involved has changed
* row_id 482 appears again, but some of the affected row_id's are different

 
#### Some other notes
* This was a **ridiculously** hard bug to recreate using test data.  I could not produce an example of this problem without resorting to generating random boolean values and generating sufficiently large datasets to increase the likelihood of getting an error.
* From my testing, the analogous pl.Expr.all() expression does not suffer from this bug.  Only the any() expression.
* I found this bug while working with boolean values that I generated using Expressions.  That is, I did not cast any values from integer to boolean.
* The error occurs even if the row_id is cast to a different type: for example, i32.  
* Complete stab in the dark: is there a missing sort of some kind when the result of the windowed any() expression is merged back into the dataset?  The fact that the windowed value is not consistent on pairs of row_id's suggests something like this.

--- a/polars/polars-lazy/src/physical_plan/expressions/alias.rs
+++ b/polars/polars-lazy/src/physical_plan/expressions/alias.rs
@@ -19,9 +19,9 @@ impl AliasExpr {
             expr,
         }
     }
-    fn finish(&self, mut input: Series) -> Result<Series> {
+    fn finish(&self, mut input: Series) -> Series {
         input.rename(&self.name);
-        Ok(input)
+        input
     }
 }
 
@@ -32,7 +32,7 @@ impl PhysicalExpr for AliasExpr {
 
     fn evaluate(&self, df: &DataFrame, state: &ExecutionState) -> Result<Series> {
         let series = self.physical_expr.evaluate(df, state)?;
-        self.finish(series)
+        Ok(self.finish(series))
     }
 
     #[allow(clippy::ptr_arg)]
@@ -43,10 +43,14 @@ impl PhysicalExpr for AliasExpr {
         state: &ExecutionState,
     ) -> Result<AggregationContext<'a>> {
         let mut ac = self.physical_expr.evaluate_on_groups(df, groups, state)?;
-        let mut s = ac.take();
-        s.rename(&self.name);
+        let s = ac.take();
+        let s = self.finish(s);
 
-        ac.with_series(self.finish(s)?, ac.is_aggregated());
+        if ac.is_literal() {
+            ac.with_literal(s);
+        } else {
+            ac.with_series(s, ac.is_aggregated());
+        }
         Ok(ac)
     }
 

--- a/polars/polars-lazy/src/physical_plan/expressions/cast.rs
+++ b/polars/polars-lazy/src/physical_plan/expressions/cast.rs
@@ -73,7 +73,13 @@ impl PhysicalExpr for CastExpr {
         let mut ac = self.input.evaluate_on_groups(df, groups, state)?;
         let s = ac.flat_naive();
         let s = self.finish(s.as_ref())?;
-        ac.with_series(s, false);
+
+        if ac.is_literal() {
+            ac.with_literal(s);
+        } else {
+            ac.with_series(s, false);
+        }
+
         Ok(ac)
     }
 

--- a/polars/polars-lazy/src/physical_plan/expressions/mod.rs
+++ b/polars/polars-lazy/src/physical_plan/expressions/mod.rs
@@ -226,6 +226,9 @@ impl<'a> AggregationContext<'a> {
     pub(crate) fn is_aggregated(&self) -> bool {
         !self.is_not_aggregated()
     }
+    pub(crate) fn is_literal(&self) -> bool {
+        matches!(self.state, AggState::Literal(_))
+    }
 
     pub(crate) fn combine_groups(&mut self, other: AggregationContext) -> &mut Self {
         if let (Cow::Borrowed(_), Cow::Owned(a)) = (&self.groups, other.groups) {
@@ -332,6 +335,11 @@ impl<'a> AggregationContext<'a> {
         self
     }
 
+    pub(crate) fn with_literal(&mut self, series: Series) -> &mut Self {
+        self.state = AggState::Literal(series);
+        self
+    }
+
     /// Update the group tuples
     pub(crate) fn with_groups(&mut self, groups: GroupsProxy) -> &mut Self {
         // In case of new groups, a series always needs to be flattened
@@ -353,6 +361,10 @@ impl<'a> AggregationContext<'a> {
                 // because this is lazy, we first must to update the groups
                 // by calling .groups()
                 self.groups();
+                assert!(
+                    self.groups.len() <= s.len(),
+                    "implementation error groups are out of bounds; please open an issue"
+                );
 
                 let out = s
                     .agg_list(&self.groups)

--- a/polars/tests/it/lazy/window_expressions.rs
+++ b/polars/tests/it/lazy/window_expressions.rs
@@ -362,3 +362,26 @@ fn test_window_exprs_any_all() -> Result<()> {
     assert!(df.frame_equal(&expected));
     Ok(())
 }
+
+#[test]
+fn test_window_naive_any() -> Result<()> {
+    let df = df![
+        "row_id" => [0, 0, 1, 1, 1],
+        "boolvar" => [true, false, true, false, false]
+    ]?;
+
+    let df = df
+        .lazy()
+        .with_column(
+            col("boolvar")
+                .sum()
+                .gt(lit(0))
+                .over([col("row_id")])
+                .alias("res"),
+        )
+        .collect()?;
+
+    let res = df.column("res")?;
+    assert_eq!(res.sum::<usize>(), Some(5));
+    Ok(())
+}

--- a/py-polars/polars/internals/lazy_functions.py
+++ b/py-polars/polars/internals/lazy_functions.py
@@ -811,7 +811,7 @@ def any(name: Union[str, List["pli.Expr"]]) -> "pli.Expr":
     """
     if isinstance(name, list):
         return fold(lit(0), lambda a, b: a | b, name).alias("any")
-    return col(name).sum() > 0
+    return col(name).any()
 
 
 def exclude(columns: Union[str, List[str]]) -> "pli.Expr":
@@ -906,7 +906,7 @@ def all(name: Optional[Union[str, List["pli.Expr"]]] = None) -> "pli.Expr":
         return col("*")
     if isinstance(name, list):
         return fold(lit(0), lambda a, b: a & b, name).alias("all")
-    return col(name).cast(bool).sum() == col(name).count()
+    return col(name).all()
 
 
 def groups(column: str) -> "pli.Expr":

--- a/py-polars/tests/test_lazy.py
+++ b/py-polars/tests/test_lazy.py
@@ -506,7 +506,7 @@ def test_all_expr() -> None:
 
 
 def test_any_expr(fruits_cars: pl.DataFrame) -> None:
-    assert fruits_cars.select(pl.any("A"))[0, 0]
+    assert fruits_cars.with_column(pl.col("A").cast(bool)).select(pl.any("A"))[0, 0]
     assert fruits_cars.select(pl.any([pl.col("A"), pl.col("B")]))[0, 0]
 
 

